junit_tests/DateTest.java

```diff
@@ -0,0 +1,353 @@
+        org.python.Object[] args2 = {org.python.types.Int.getInt(2020), org.python.types.Int.getInt(8), org.python.types.Int.getInt(14)};
+        org.python.stdlib.datetime.Date sameDate = new org.python.stdlib.datetime.Date(args2, Collections.emptyMap());
+        org.python.Object[] args3 = {org.python.types.Int.getInt(2020), org.python.types.Int.getInt(8), org.python.types.Int.getInt(16)};
+        org.python.stdlib.datetime.Date greaterDate = new org.python.stdlib.datetime.Date(args3, Collections.emptyMap());
+        org.python.Object[] args4 = {org.python.types.Int.getInt(2019), org.python.types.Int.getInt(8), org.python.types.Int.getInt(16)};
+        org.python.stdlib.datetime.Date smallerDate = new org.python.stdlib.datetime.Date(args4, Collections.emptyMap());
+        org.python.types.Str notADate = new org.python.types.Str("I am not a Date");
+
+        List<Map.Entry<String, Function<org.python.Object, org.python.Object>>> functions = new ArrayList<>();
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Eq", testDate::__eq__));
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Ne", testDate::__ne__));
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Le", testDate::__le__));
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Lt", testDate::__lt__));
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Ge", testDate::__ge__));
+        functions.add(new AbstractMap.SimpleImmutableEntry<>("Gt", testDate::__gt__));
+
+        boolean[] resultsSameDate = {true, false, true, false, true, false};
+        boolean[] resultsGreaterDate = {false, true, true, true, false, false};
+        boolean[] resultsSmallerDate = {false, true, false, false, true, true};
+
+        List<DynamicTest> tests = new ArrayList<>(functions.size());
+        for (int i = 0; i < functions.size(); i++) {
+            var entry = functions.get(i);
+            var funcName = entry.getKey();
+            var function = entry.getValue();
+            int j = i;
+
+            tests.add(DynamicTest.dynamicTest(
+                "test" + funcName + "_Same",
+                () -> assertEquals(resultsSameDate[j], ((org.python.types.Bool) function.apply(sameDate)).value)));
+            tests.add(DynamicTest.dynamicTest(
+                "test" + funcName + "_Greater",
+                () -> assertEquals(resultsGreaterDate[j], ((org.python.types.Bool) function.apply(greaterDate)).value)));
+            tests.add(DynamicTest.dynamicTest(
+                "test" + funcName + "_Smaller",
+                () -> assertEquals(resultsSmallerDate[j], ((org.python.types.Bool) function.apply(smallerDate)).value)));
+            tests.add(DynamicTest.dynamicTest(
+                "test" + funcName + "_Invalid",
+                () ->assertEquals(org.python.types.NotImplementedType.NOT_IMPLEMENTED, function.apply(notADate))));
```

Comment:
This is actually a tricky case, because the operators don't behave the same as the functions. Consider this:
```python
>>> from datetime import date
>>> date(1, 2, 3) < 42
TypeError: '<' not supported between instances of 'datetime.date' and 'int'
>>> date(1, 2, 3).__lt__(42)
NotImplemented
```
This is because the operator behavior is a bit more complicated. If `a.__lt__(b)` returns NotImplemented, Python will also try `b.__ge__(a)` and only if that as well returns NotImplemented fail with the TypeError.

I don't know if it is handled the same way in VOC, but e.g. the comparison functions in `org.python.types.Bool` return the NotImplemented object.

That this is also the expected behavior for the date comparison methods can be seen by looking at the [Python implementation](https://github.com/python/cpython/blob/851811f5772c43f72f445e2ce1ac3ea9da951ae3/Lib/datetime.py#L995).

Reply:
Since you're returning the correct error when using the function directly, I guess that is the way to do it, so no more comments on this. :-)